src/tree/mod.rs

```diff
@@ -1,103 +1,363 @@
+
         Self {
-            tree_data: Arc::new(RwLock::new(TreeData::new(
-                tree,
-                tree_history_size,
-            ))),
-            tree_updater: Arc::new(TreeUpdater::new(
-                address,
-                creation_block,
-                window_size,
-                middleware,
-            )),
-            synced: Arc::new(AtomicBool::new(false)),
+            identity_tree: Arc::new(RwLock::new(identity_tree)),
+            canonical_tree_manager,
+            bridged_tree_manager,
+            chain_state: Arc::new(RwLock::new(HashMap::new())),
+            synced: AtomicBool::new(false),
         }
     }
 
-    /// Spawns a task that continually syncs the `TreeData` to the state at the chain head.
-    #[instrument(skip(self))]
-    pub fn spawn(&self) -> JoinHandle<Result<(), TreeAvailabilityError<M>>> {
-        let tree_data = self.tree_data.clone();
-        let tree_updater = self.tree_updater.clone();
+    pub async fn spawn(&self) -> eyre::Result<Vec<JoinHandle<()>>> {
+        let start_time = Instant::now();
 
-        tracing::info!("Spawning thread to sync tree");
-        let synced = self.synced.clone();
+        tracing::info!("Syncing to head");
+        self.sync_to_head().await?;
+        tracing::info!("Synced to head in {:?} seconds", start_time.elapsed());
 
-        tokio::spawn(async move {
-            let start = tokio::time::Instant::now();
-            tree_updater.sync_to_head(&tree_data).await?;
-            let sync_time = start.elapsed();
+        let (leaf_updates_tx, leaf_updates_rx) =
+            tokio::sync::mpsc::channel(100);
+        let (bridged_root_tx, bridged_root_rx) =
+            tokio::sync::mpsc::channel(100);
+
+        // Spawn the tree managers for the canonical and bridged trees
+        let mut handles = vec![];
+        handles.push(self.canonical_tree_manager.spawn(leaf_updates_tx));
+
+        //TODO: handle if there are no bridged roots to spawn, otherwise the channel will close
+        for bridged_tree in self.bridged_tree_manager.iter() {
+            handles.push(bridged_tree.spawn(bridged_root_tx.clone()));
+        }
+
+        handles
+            .push(self.handle_tree_updates(leaf_updates_rx, bridged_root_rx));
+
+        Ok(handles)
+    }
 
-            tracing::info!(?sync_time, "WorldTree synced to chain head");
-            synced.store(true, Ordering::Relaxed);
+    fn handle_tree_updates(
+        &self,
+        mut leaf_updates_rx: Receiver<(Root, LeafUpdates)>,
+        mut bridged_root_rx: Receiver<(u64, Hash)>,
+    ) -> JoinHandle<()> {
+        let identity_tree = self.identity_tree.clone();
+        let chain_state = self.chain_state.clone();
+        let canonical_chain_id = self.canonical_tree_manager.chain_id;
 
+        tokio::spawn(async move {
             loop {
-                tree_updater.sync_to_head(&tree_data).await?;
+                async {
+                    tokio::select! {
```

Comment:
Is there a good reason why we have to select here? Could we just spawn 2 tasks - one for canonical and the other for bridged channel proceesing?

Reply:
I'm not super opposed to the idea btw. I just think maybe that way we could reduce nesting and simplify this/these loops

Reply:
Yeah, Ill update to spawn two tasks. This will be cleaner.